blog/_posts/2020-05-16-next-few-years.adoc

```diff
@@ -0,0 +1,193 @@
+
+The amount of programming work on rust-analyzer side is relatively small here:
+we need to fix various protocol conformance issues,
+clean up various defaults to be less experimental,
+write documentation which doesn't require a lot of enthusiasm to understand, etc.
+The amount of org stuff is much bigger -- we need to package rust-analyzer with rustup, merge RLS and rust-analyzer VS Code extension, figure out repository structure, etc.
+
+Separately, I want to make sure that rust-analyzer is usable inside large non-Cargo based monorepos.
+We have some initial support for this already, but there's a bunch of details we need to iron out.
+
+== Dangers of Accidental Architecture
+
+The main danger I see is that rust-analyzer can ossify, as it is today.
+This would be bad, because, although current rust-analyzer architecture is right in broad strokes, a lot of important and hard-to-change details are wrong.
+After we push rust-analyzer to the general public, we should focus on boring implementation & design work, with relatively few shiny gifs and a lot of foundational work for the decade.
+
+== Bringing Chalk to Rustc
+
+rust-analyzer has been using chalk as its trait solver for a long time now.
+It would be good to finish the work, and integrate it into rustc as well, ~~and give people their GATs~~.
+
+== Single Parser and Syntax Tree
+
+We should share the parser between rustc and rust-analyzer already.
+Parsing is one of the most interesting bits of the compiler, from the IDE point of view.
+By transitioning rustc to a lossless syntax we'll cross the most important barrier, and it will be a downhill road from that point on.
+The design space here I think is well-understood, but the amount of work to do is large.
+At some point, I should take a break from actively hacking on rust-analyzer and focus on sharing the parser.
+
+== Virtual File System
+
+The most fundamental data structure in rust-analyzer, even more fundamental than a syntax tree, is VFS.
+It serves two dual goals:
+
+* providing consistent immutable snapshots of the file system,
+* applying transactional changes to the state of the file system.
+
+This abstraction is the boundary between pure-functional universe of rust-analyzer, and the messiness of external world.
+It needs to bridge case-insensitive file systems, symlinks and cycles to a simpler model of "tree with utf8 paths" we want inside.
+Additionally it should work with non-path files: there are use-cases where we want to do analysis of Rust code, which doesn't really reside on the file system.
+
+One specific aspect I am struggling with is dynamism.
+On the one hand, it seems that a good design is to require to specify the set of files in VFS upfront, as a set of globs.
+This is important because, to properly register file watchers without loosing updates, it's important to crawl the file-system eagerly.
```

Comment:
```suggestion
This is important because, to properly register file watchers without losing updates, it's important to crawl the file-system eagerly.
```